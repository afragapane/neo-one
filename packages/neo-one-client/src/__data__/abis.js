/* @flow */
import type {
  ABIEvent,
  ABIFunction,
  ABIParameter,
  ABIReturn,
  ABI,
} from '../types';

export const returns = {
  Signature: { type: 'Signature' },
  Boolean: { type: 'Boolean' },
  Hash160: { type: 'Hash160' },
  Hash256: { type: 'Hash256' },
  ByteArray: { type: 'ByteArray' },
  PublicKey: { type: 'PublicKey' },
  String: { type: 'String' },
  InteropInterface: { type: 'InteropInterface' },
  Void: { type: 'Void' },
  Integer: { type: 'Integer', decimals: 7 },
  Array: { type: 'Array', value: { type: 'String' } },
};

export const parameters = {
  Signature: { name: 'signatureName', type: 'Signature' },
  Boolean: { name: 'booleanName', type: 'Boolean' },
  Integer: { name: 'intName', type: 'Integer', decimals: 7 },
  Hash160: { name: 'hash160Name', type: 'Hash160' },
  Hash256: { name: 'hash256Name', type: 'Hash256' },
  ByteArray: { name: 'byteName', type: 'ByteArray' },
  PublicKey: { name: 'pubkeyName', type: 'PublicKey' },
  String: { name: 'stringName', type: 'String' },
  Array: {
    name: 'arrayName',
    type: 'Array',
    value: {
      name: 'stringName',
      type: 'String',
    },
  },
  InteropInterface: { name: 'interopName', type: 'InteropInterface' },
  Void: { name: 'voidName', type: 'Void' },
};

export const abiFunction = (
  constant: boolean = false,
  params: Array<ABIParameter> = [parameters.String],
  verify: boolean = false,
  name: string = 'funcName',
  returnType: ABIReturn = returns.Void,
): ABIFunction => ({
  name,
  constant,
  parameters: params,
  verify,
  returnType,
});

export const abiEvent = (
  name: string = 'eventName',
  params: Array<ABIParameter> = [parameters.String],
): ABIEvent => ({
  name,
  parameters: params,
});

export const abi = (
  functions: Array<ABIFunction> = [abiFunction()],
  events?: Array<ABIEvent> = [],
): ABI => ({
  functions,
  events,
});
