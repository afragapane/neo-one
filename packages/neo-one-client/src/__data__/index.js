/* @flow */
export { default as keys } from './keys';
export { default as transactions } from './transactions';
