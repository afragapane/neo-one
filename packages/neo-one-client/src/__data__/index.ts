export * from './keys';
export * from './data';
export * from './transactions';
export * from './factory';
