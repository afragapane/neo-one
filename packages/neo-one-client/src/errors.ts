import BigNumber from 'bignumber.js';
import { CustomError } from '@neo-one/utils';
import { ContractParameter, ContractParameterType } from './types';

export class InvalidContractParameterError extends CustomError {
  public readonly parameter: ContractParameter;
  public readonly expected: ContractParameterType[];
  public readonly code: string;

  constructor(parameter: ContractParameter, expected: ContractParameterType[]) {
    super(
      `Expected one of ${JSON.stringify(expected)} ` +
        `ContractParameterTypes, found ${parameter.type}`,
    );

    this.parameter = parameter;
    this.expected = expected;
    this.code = 'INVALID_CONTRACT_PARAMETER';
  }
}

export class InvalidArgumentError extends CustomError {
  public readonly code: string;

  constructor(message: string) {
    super(message);
    this.code = 'INVALID_ARGUMENT';
  }
}

export class InvalidNamedArgumentError extends InvalidArgumentError {
  public readonly code: string;

  constructor(name: string, argument: {}) {
    super(`Invalid argument for ${name}: ${String(argument)}`);
    this.code = 'INVALID_NAMED_ARGUMENT';
  }
}

export class InvocationCallError extends CustomError {
  public readonly code: string;

  constructor(message: string) {
    super(message);
    this.code = 'INVOCATION_CALL';
  }
}

export class InvalidEventError extends CustomError {
  public readonly code: string;

  constructor(message: string) {
    super(message);
    this.code = 'INVALID_EVENT';
  }
}

export class InsufficientFundsError extends CustomError {
  public readonly code: string;

  constructor(total: BigNumber, expected: BigNumber) {
    super(`Found ${total.toString()} funds, required: ${expected.toString()}.`);
    this.code = 'INSUFFICIENT_FUNDS';
  }
}

export class NothingToClaimError extends CustomError {
  public readonly code: string;

  constructor() {
    super('Nothing to claim.');
    this.code = 'NOTHING_TO_CLAIM';
  }
}

export class NothingToIssueError extends CustomError {
  public readonly code: string;

  constructor() {
    super('Nothing to issue.');
    this.code = 'NOTHING_TO_ISSUE';
  }
}

export class NothingToTransferError extends CustomError {
  public readonly code: string;

  constructor() {
    super('Nothing to transfer.');
    this.code = 'NOTHING_TO_TRANSFER';
  }
}

export class NoAccountError extends CustomError {
  public readonly code: string;

  constructor() {
    super('No account exists.');
    this.code = 'NO_ACCOUNT';
  }
}

export class NoContractDeployedError extends CustomError {
  public readonly code: string;

  constructor(networkType: string) {
    super(`Contract has not been deployed to network ${networkType}`);
    this.code = 'NO_CONTRACT_DEPLOYED';
  }
}

export class InvalidTransactionError extends CustomError {
  public readonly code: string;

  constructor(message: string) {
    super(message);
    this.code = 'INVALID_TRANSACTION';
  }
}

export class InvokeError extends CustomError {
  public readonly code: string;

  constructor(message: string) {
    super(message);
    this.code = 'INVOKE';
  }
}

export class UnknownBlockError extends CustomError {
  public readonly code: string;

  constructor() {
    super('Unknown block');
    this.code = 'UNKNOWN_BLOCK';
  }
}

export class RelayTransactionError extends CustomError {
  public readonly code: string;

  constructor(message: string) {
    super(message);
    this.code = 'RELAY_TRANSACTION';
  }
}

export class UnknownNetworkError extends CustomError {
  public readonly code: string;

  constructor(name: string) {
    super(`Unknown network ${name}`);
    this.code = 'UNKNOWN_NETWORK';
  }
}

export class UnknownAccountError extends CustomError {
  public readonly code: string;

  constructor(address: string) {
    super(`Unknown account ${address}`);
    this.code = 'UNKNOWN_ACCOUNT';
  }
}

export class LockedAccountError extends CustomError {
  public readonly code: string;

  constructor(address: string) {
    super(`Account ${address} is locked`);
    this.code = 'LOCKED_ACCOUNT';
  }
}

export class PasswordRequiredError extends CustomError {
  public readonly code: string;

  constructor() {
    super('A password is required when creating accounts on the MainNet.');
    this.code = 'PASSWORD_REQUIRED';
  }
}
