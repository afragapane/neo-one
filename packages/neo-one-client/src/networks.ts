// tslint:disable-next-line export-name
export const MAIN = 'main';
