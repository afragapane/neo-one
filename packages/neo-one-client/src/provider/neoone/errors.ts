import { CustomError } from '@neo-one/utils';

export interface JSONRPCErrorResponse {
  code: number;
  message: string;
  data?: any;
}

export class JSONRPCError extends CustomError {
  public readonly responseError: JSONRPCErrorResponse;
  public readonly code: string;

  constructor(responseError: JSONRPCErrorResponse) {
    super(responseError.message);
    this.responseError = responseError;
    this.code = 'JSON_RPC';
  }
}

export class InvalidRPCResponseError extends CustomError {
  public readonly code: string;

  constructor() {
    super('Did not receive valid rpc response');
    this.code = 'INVALID_RPC_RESPONSE';
  }
}

export class HTTPError extends CustomError {
  public readonly status: number;
  public readonly text: string | null;
  public readonly code: string;

  constructor(status: number, text: string | null) {
    let message = `HTTP Error ${status}`;
    if (text != null) {
      message = `${message}: ${text}`;
    }
    super(message);
    this.status = status;
    this.text = text;
    this.code = 'HTTP';
  }
}

export class MissingTransactionDataError extends CustomError {
  public readonly code: string = 'MISSING_TRANSACTION_DATA';

  constructor(hash: string) {
    super(`Missing transaction data for transaction ${hash}`);
  }
}
