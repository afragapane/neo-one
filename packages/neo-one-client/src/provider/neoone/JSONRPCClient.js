/* @flow */
import {
  type AccountJSON,
  type AssetJSON,
  type BlockJSON,
  type ContractJSON,
  type InputJSON,
  type InvocationDataJSON,
  type InvocationResultJSON,
  type NetworkSettingsJSON,
  type OutputJSON,
  type StorageItemJSON,
  type TransactionJSON,
  type TransactionReceiptJSON,
  type ValidatorJSON,
} from '@neo-one/client-core';
import BigNumber from 'bignumber.js';

import {
  type AddressString,
  type BufferString,
  type GetOptions,
  type Hash160String,
  type Hash256String,
  type Peer,
  type Options,
} from '../../types';
import { type JSONRPCProvider } from './JSONRPCProvider';
import { RelayTransactionError } from '../../errors';

export default class JSONRPCClient {
  _provider: JSONRPCProvider;

  constructor(provider: JSONRPCProvider) {
    this._provider = provider;
  }

  getAccount(hash: AddressString): Promise<AccountJSON> {
    return this._provider.request({
      method: 'getaccountstate',
      params: [hash],
    });
  }

  getAsset(hash: Hash256String): Promise<AssetJSON> {
    return this._provider.request({
      method: 'getassetstate',
      params: [hash],
    });
  }

  getBlock(
    hashOrIndex: Hash256String | number,
    options?: GetOptions,
  ): Promise<BlockJSON> {
    const { timeoutMS } = options || {};
    return this._provider.request({
      method: 'getblock',
      params: [hashOrIndex, 1],
      watchTimeoutMS: timeoutMS,
    });
  }

  getBestBlockHash(): Promise<string> {
    return this._provider.request({ method: 'getbestblockhash' });
  }

  getBlockCount(): Promise<number> {
    return this._provider.request({ method: 'getblockcount' });
  }

  getContract(hash: Hash160String): Promise<ContractJSON> {
    return this._provider.request({
      method: 'getcontractstate',
      params: [hash],
    });
  }

  getMemPool(): Promise<Array<string>> {
    return this._provider.request({ method: 'getrawmempool' });
  }

  getTransaction(hash: Hash256String): Promise<TransactionJSON> {
    return this._provider.request({
      method: 'getrawtransaction',
      params: [hash, 1],
    });
  }

  getStorageItem(
    hash: Hash160String,
    key: BufferString,
  ): Promise<StorageItemJSON> {
    return this._provider
      .request({
        method: 'getstorage',
        params: [hash, key],
      })
      .then(value => ({ hash, key, value }));
  }

  getUnspentOutput(input: InputJSON): Promise<?OutputJSON> {
    return this._provider.request({
      method: 'gettxout',
      params: [input.txid, input.vout],
    });
  }

  testInvokeRaw(script: BufferString): Promise<InvocationResultJSON> {
    return this._provider.request({
      method: 'invokescript',
      params: [script],
    });
  }

  sendTransactionRaw(value: BufferString): Promise<void> {
    return this._provider
      .request({
        method: 'sendrawtransaction',
        params: [value],
      })
      .then(result => {
        if (!result) {
          throw new RelayTransactionError('Relay transaction failed.');
        }

        return result;
      });
  }

  relayTransaction(value: BufferString): Promise<TransactionJSON> {
    return this._provider
      .request({
        method: 'relaytransaction',
        params: [value],
      })
      .catch(error => {
        if (error.code === 'JSON_RPC' && error.responseError.code === -110) {
          throw new RelayTransactionError(error.responseError.message);
        }

        throw error;
      });
  }

  getOutput(input: InputJSON): Promise<OutputJSON> {
    return this._provider.request({
      method: 'getoutput',
      params: [input.txid, input.vout],
    });
  }

  getClaimAmount(input: InputJSON): Promise<BigNumber> {
    return this._provider
      .request({
        method: 'getclaimamount',
        params: [input.txid, input.vout],
      })
      .then(res => new BigNumber(res));
  }

  getAllStorage(hash: Hash160String): Promise<Array<StorageItemJSON>> {
    return this._provider.request({
      method: 'getallstorage',
      params: [hash],
    });
  }

  testInvocation(value: BufferString): Promise<InvocationResultJSON> {
    return this._provider.request({
      method: 'testinvocation',
      params: [value],
    });
  }

  getTransactionReceipt(
    hash: Hash256String,
    options?: GetOptions,
  ): Promise<TransactionReceiptJSON> {
    const { timeoutMS } = options || {};
    return this._provider.request({
      method: 'gettransactionreceipt',
      params: [hash],
      watchTimeoutMS: timeoutMS,
    });
  }

  getInvocationData(hash: Hash256String): Promise<InvocationDataJSON> {
    return this._provider.request({
      method: 'getinvocationdata',
      params: [hash],
    });
  }

  getValidators(): Promise<Array<ValidatorJSON>> {
    return this._provider.request({
      method: 'getvalidators',
    });
  }

  getConnectedPeers(): Promise<Array<Peer>> {
    return this._provider
      .request({
        method: 'getpeers',
      })
      .then(result => result.connected);
  }

  getNetworkSettings(): Promise<NetworkSettingsJSON> {
    return this._provider.request({
      method: 'getnetworksettings',
    });
  }

  runConsensusNow(): Promise<void> {
    return this._provider.request({
      method: 'runconsensusnow',
    });
  }

  updateSettings(options: Options): Promise<void> {
    return this._provider.request({
      method: 'updatesettings',
      params: [options],
    });
  }
}
