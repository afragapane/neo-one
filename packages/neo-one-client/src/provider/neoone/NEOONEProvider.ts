import { Monitor } from '@neo-one/monitor';
import BigNumber from 'bignumber.js';
import { BehaviorSubject, Observable } from 'rxjs';
import { UnknownNetworkError } from '../../errors';
import * as networkConfigs from '../../networks';
import {
  AddressString,
  GetOptions,
  Hash256String,
  Input,
  NetworkSettings,
  NetworkType,
  RawInvocationData,
  RawInvocationResult,
  Transaction,
  TransactionReceipt,
  UnspentOutput,
} from '../../types';
import { NEOONEDataProvider } from './NEOONEDataProvider';

export interface NEOONEProviderOptions {
  network: NetworkType;
  rpcURL: string;
}

export class NEOONEProvider {
  public readonly networks$: Observable<NetworkType[]>;
  private readonly networksInternal$: BehaviorSubject<NetworkType[]>;
  private readonly providers: { [type: string]: NEOONEDataProvider };

  constructor(
    input: {
      mainRPCURL?: string;
      testRPCURL?: string;
      options?: NEOONEProviderOptions[];
    } = {},
  ) {
    const {
      mainRPCURL: mainRPCURLIn,
      testRPCURL: testRPCURLIn,
      options,
    } = input;
    this.networksInternal$ = new BehaviorSubject([] as NetworkType[]);
    this.networks$ = this.networksInternal$;
    this.providers = {};

    let hasMain = false;
    let hasTest = false;
    const networks = (options || []).map(({ network, rpcURL }) => {
      if (network === networkConfigs.MAIN) {
        hasMain = true;
      }

      if (network === networkConfigs.TEST) {
        hasTest = true;
      }

      this.providers[network] = new NEOONEDataProvider({
        network,
        rpcURL,
      });

      return network;
    });

    if (!hasMain) {
      const mainRPCURL =
        mainRPCURLIn == null ? networkConfigs.MAIN_URL : mainRPCURLIn;
      this.providers.main = new NEOONEDataProvider({
        network: networkConfigs.MAIN,
        rpcURL: mainRPCURL,
      });

      networks.push(networkConfigs.MAIN);
    }

    if (!hasTest) {
      const testRPCURL =
        testRPCURLIn == null ? networkConfigs.TEST_URL : testRPCURLIn;
      this.providers.test = new NEOONEDataProvider({
        network: networkConfigs.TEST,
        rpcURL: testRPCURL,
      });

      networks.push(networkConfigs.TEST);
    }

    this.networksInternal$.next(networks);
  }

  public getNetworks(): NetworkType[] {
    return this.networksInternal$.getValue();
  }

  public addNetwork({
    network,
    rpcURL,
  }: {
    network: NetworkType;
    rpcURL: string;
  }): void {
    this.providers[network] = new NEOONEDataProvider({ network, rpcURL });
    const networks = this.networksInternal$.value.filter(
      (net) => network !== net,
    );
    networks.push(network);
    this.networksInternal$.next(networks);
  }

  public getUnclaimed(
    network: NetworkType,
    address: AddressString,
    monitor?: Monitor,
  ): Promise<{ unclaimed: Input[]; amount: BigNumber }> {
    return this.getProvider(network).getUnclaimed(address, monitor);
  }

  public getUnspentOutputs(
    network: NetworkType,
    address: AddressString,
    monitor?: Monitor,
  ): Promise<UnspentOutput[]> {
    return this.getProvider(network).getUnspentOutputs(address, monitor);
  }

  public relayTransaction(
    network: NetworkType,
    transaction: string,
    monitor?: Monitor,
  ): Promise<Transaction> {
    return this.getProvider(network).relayTransaction(transaction, monitor);
  }

  public getTransactionReceipt(
    network: NetworkType,
    hash: Hash256String,
    options?: GetOptions,
  ): Promise<TransactionReceipt> {
    return this.getProvider(network).getTransactionReceipt(hash, options);
  }

  public getInvocationData(
    network: NetworkType,
    hash: Hash256String,
    monitor?: Monitor,
  ): Promise<RawInvocationData> {
    return this.getProvider(network).getInvocationData(hash, monitor);
  }

  public testInvoke(
    network: NetworkType,
    transaction: string,
    monitor?: Monitor,
  ): Promise<RawInvocationResult> {
    return this.getProvider(network).testInvoke(transaction, monitor);
  }

  public getNetworkSettings(
    network: NetworkType,
    monitor?: Monitor,
  ): Promise<NetworkSettings> {
    return this.getProvider(network).getNetworkSettings(monitor);
  }

  public read(network: NetworkType): NEOONEDataProvider {
    return this.getProvider(network);
  }

  private getProvider(network: NetworkType): NEOONEDataProvider {
    const provider = this.providers[network];
    if (provider == null) {
      throw new UnknownNetworkError(network);
    }

    return provider;
  }
}
