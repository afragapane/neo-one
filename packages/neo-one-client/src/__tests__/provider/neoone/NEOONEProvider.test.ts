import { keys, transactions } from '../../../__data__';
import { UnknownNetworkError } from '../../../errors';
import * as networkConfigs from '../../../networks';
import { NEOONEProvider } from '../../../provider/neoone/NEOONEProvider';

const MAIN_URL = 'https://neotracker.io/rpc';

describe('NEOONEProvider', () => {
  const network = 'net';
  const expected = '0';
  const options = [{ network, rpcURL: 'rpc' }];

  let provider = new NEOONEProvider(options);
  beforeEach(() => {
    provider = new NEOONEProvider(options);
  });

  test('NEOONEProvider constructor with no options', async () => {
    const testProvider = new NEOONEProvider();
    const result = await testProvider.getNetworks();
    expect(result).toEqual([]);
  });

  test('NEOONEProvider constructor with main & test', async () => {
    const testOptions = [{ network: networkConfigs.MAIN, rpcURL: MAIN_URL }];

    const testExpected = testOptions.map((option) => option.network);

    const testProvider = new NEOONEProvider(testOptions);
    const result = await testProvider.getNetworks();
    expect(result).toEqual(testExpected);
  });

  test('getNetworks', async () => {
    // @ts-ignore
    provider.networks$.getValue = jest.fn(() => '0');

    const result = provider.getNetworks();
    expect(result).toEqual(expected);
  });

  test('addNetwork new network', async () => {
    const newNetwork = 'newNet';
    provider.addNetwork({ network: newNetwork, rpcURL: 'rpc' });

    const result = await provider.getNetworks();
    expect(result).toEqual([network, newNetwork]);
  });

  test('addNetwork existing network', async () => {
    provider.addNetwork({ network, rpcURL: 'rpc' });

    const result = await provider.getNetworks();
    expect(result).toEqual([network]);

    provider.addNetwork({ network, rpcURL: 'rpc' });
    const newResult = await provider.getNetworks();
    expect(newResult).toEqual([network]);
  });

  test('read', () => {
    const result = provider.read(network);
    expect(result).toMatchSnapshot();
  });

  test('_getProvider throws error for unknown network', () => {
    const fakeNet = 'fake';
    function testError() {
      // @ts-ignore
      provider.getProvider(fakeNet);
    }
    expect(testError).toThrow(new UnknownNetworkError(fakeNet) as any);
  });

  const testCases = [
    {
      method: 'getUnclaimed',
      args: [network, keys[0].address],
    },

    {
      method: 'getUnspentOutputs',
      args: [network, keys[0].address],
    },

    {
      method: 'relayTransaction',
      args: [network, ''],
    },

    {
      method: 'getTransactionReceipt',
      args: [network, transactions.register.hash],
    },

    {
      method: 'getInvocationData',
      args: [network, transactions.register.hash],
    },

    {
      method: 'testInvoke',
      args: [network, ''],
    },

    {
      method: 'getNetworkSettings',
      args: [network],
    },
  ];

  for (const testCase of testCases) {
    const { method, args } = testCase;
    // eslint-disable-next-line
    test(method, async () => {
      // @ts-ignore
      provider.mutableProviders[network][method] = jest.fn(() => expected);
      // @ts-ignore
      const result = await provider[method](...args);
      expect(result).toEqual(expected);
    });
  }
});
