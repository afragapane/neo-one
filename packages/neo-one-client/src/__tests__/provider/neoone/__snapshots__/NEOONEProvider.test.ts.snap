// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NEOONEProvider getBlockCount - unknown network 1`] = `[Error: Unknown network unknown]`;
